### ISSUE TITLE
"'NoneType' object has no attribute '_fetch_cloud_tpu_metadata'" when using TPU

### ISSUE THREAD

**Issue by w4-sjcho** [tensorflow/tensorflow#30869]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.0
- GPU model and memory: Cloud TPU v2

**Describe the current behavior**
I've been using T2T training script using cloud TPU. After upgrading to TF 1.14.0 from TF 1.13.1, I'm getting the error below:

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/app/poodle/asr/t2t/t2t_train_venv.image.binary.runfiles/skelterlabs/poodle/asr/t2t/py-pkg/tensorflow/python/tpu/preempted_hook.py", line 86, in run
    response = self._cluster._fetch_cloud_tpu_metadata()  # pylint: disable=protected-access
AttributeError: 'NoneType' object has no attribute '_fetch_cloud_tpu_metadata'
```

The training script continues w/o any further issue despite the error.

**Describe the expected behavior**
No error.

**Code to reproduce the issue**
Run https://github.com/tensorflow/tensor2tensor#speech-recognition problems using transformer model on TPUs.

**Other info / logs**
N/A

**Comment by HeimingX** [comment]
I also meet this problem, have you solved it?

**Comment by jvishnuvardhan** [comment]
@w4-sjcho @HeimingX Do you have issues with this [colab](https://colab.sandbox.google.com/github/tensorflow/tensor2tensor/blob/master/tensor2tensor/notebooks/hello_t2t.ipynb#scrollTo=wfF8_cW-OXPN). Please let us know the location/item which is breaking. Thanks!

**Comment by cwbeitel** [comment]
@jvishnuvardhan @jhseu Also seeing this with tensorflow 1.14 and t2t 1.14, happy to help. This is running in batch on ctpu v3 with tf 1.14 runtime (in graph mode).

As you can see the error above `self._cluster._fetch_cloud_tpu_metadata()` is failing because self._cluster is None; this is being passed to CloudTPUPreemptedHook [in tpu_estimator.py](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py#L3237).

See there has recently been a [change to tpu_estimator.py](https://github.com/tensorflow/estimator/commit/6a8b06637a73d070ca5ccbeeb5fdc280c8aa9994) that modifies:

```python
         if tpu_cluster_resolver.is_running_in_gce():
            hooks.extend(
                [preempted_hook.CloudTPUPreemptedHook(self._config.cluster)])
```
to

```python
         if _check_add_preemption_hook(self._config.cluster):
            hooks.extend(
                [preempted_hook.CloudTPUPreemptedHook(self._config.cluster)])
```

where the latter wraps the former to include a check for the TPUClusterResolver object not being None, here:

```python
def _check_add_preemption_hook(cluster):
  return (tpu_cluster_resolver.is_running_in_gce() and
          cluster and
          isinstance(cluster, tpu_cluster_resolver.TPUClusterResolver) and
          cluster._should_resolve)
```

which seems like one solution (i.e. avoid adding the problematic hook if we don't have a cluster resolver object).

Another would be to ensure the cluster resolver gets initialized when the run config is constructed in t2t trainer_lib.py, [here](https://github.com/tensorflow/tensor2tensor/blob/2036ffe309b86bda367b1e687fafb114534500f9/tensor2tensor/utils/trainer_lib.py#L232); looking at the logic there the portion initializing the cluster resolver won't be reached when using TPUs via GKE because KUBE_GOOGLE_CLOUD_TPU_ENDPOINTS will be set.

Idk if it will break something else but I'm just going to try moving the initialization of the cluster resolver outside of the logic about whether master and KUBE_GOOGLE_CLOUD_TPU_ENDPOINTS are set. Or re-build tf with the modification to tpu_estimator.py mentioned above. The tpu runtime tf version would still be stock 1.14 but afaict that pertains to ops not session hooks.

**Comment by cwbeitel** [comment]
@jvishnuvardhan @jhseu @w4-sjcho @HeimingX So it does look like there's a side effect in moving the cluster resolver instantiation from its current place but [patching in](https://gist.github.com/cwbeitel/2ab6bb9c5c418d390bedb23b8c959c0b#file-patch_tpu_estimator-py) the `_check_add_preemption_hook` change mentioned above gets rid of the error and training appears to proceed normally.

**Comment by jhseu** [comment]
Yeah, this'll be fixed in TF 1.15. You can also pass the cluster to the RunConfig instead of passing master directly and it should work.
